src/io/flutter/sdk/FlutterSdk.java

```diff
@@ -141,14 +137,36 @@ public FlutterCommand flutterRun(@NotNull PubRoot root, @NotNull VirtualFile mai
     args.addAll(asList(additionalArgs));
 
     // Make the path to main relative (to make the command line prettier).
-    assert main.getPath().startsWith(root.getPath());
-    assert !root.getPath().endsWith("/");
-    final String mainPath = main.getPath().substring(root.getPath().length() + 1);
+    final String mainPath = root.getRelativePath(main);
+    if (mainPath == null) {
+      throw new IllegalArgumentException("main isn't within the pub root: " + main.getPath());
+    }
     args.add(FileUtil.toSystemDependentName(mainPath));
 
     return new FlutterCommand(this, root.getRoot(), FlutterCommand.Type.RUN, args.toArray(new String[]{}));
   }
 
+  public FlutterCommand flutterTest(@NotNull PubRoot root, @NotNull VirtualFile main, @Nullable FlutterDevice device) {
+
+    // We don't have machine mode yet, so just run it normally and show the output in the console.
+    final List<String> args = new ArrayList<>();
+    if (FlutterInitializer.isVerboseLogging()) {
+      args.add("--verbose");
+    }
+    if (device != null) {
+      args.add("--device-id=" + device.deviceId());
```

Comment:
does `flutter test` take a device ID? doesn't is just run test against sky_shell?